sycl/doc/extensions/proposed/sycl_ext_oneapi_barrier.asciidoc

```diff
@@ -0,0 +1,260 @@
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== `barrier` class
+
+The `sycl::ext::oneapi::experimental::barrier` class has the same interface as
+the `std::barrier` class from {cpp}20, with an additional `Scope` template
+parameter denoting a memory scope broad enough to contain all threads of
+execution that are able to use the barrier.
+
+The meaning of the `Scope` template parameter is similar to the meaning of the
+scope associated with a `sycl::atomic_ref`, and is orthogonal to the memory
+space in which the barrier object itself is allocated. The `Scope` template
+is effectively a promise that the barrier will only be used to synchronize
+threads of execution within the same scope, which may allow an implementation
+to improve performance. If the scope associated with a barrier is
+insufficiently broad to contain all threads of execution that attempt to use
+the barrier, behavior is undefined.
+
+[NOTE]
+====
+The threads of execution that are synchronized by this barrier may be a
+combination of host threads and device threads (i.e. work-items), if the
+barrier is visible to all threads of execution and has a sufficiently broad
+scope.
+====
+
+This extension specification provides a brief overview of the functionality of
+the `barrier` class. If there are any differences in the definitions of this
+functionality, the wording in the {cpp}20 specification takes precedence.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  template <sycl::memory_scope Scope, class CompletionFunction = __unspecified__>
+  class barrier {
+  public:
+    using arrival_token = __unspecified__;
+
+    static constexpr ptrdiff_t max() noexcept;
+
+    constexpr explicit barrier(ptrdiff_t expected,
+                               CompletionFunction f = CompletionFunction());
+    ~barrier();
+
+    barrier(const barrier&) = delete;
+    barrier& operator=(const barrier&) = delete;
+
+    [[nodiscard]] arrival_token arrive(ptrdiff_t update = 1);
+    void wait(arrival_token&& arrival) const;
+
+    void arrive_and_wait();
+    void arrive_and_drop();
+  };
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+[source,c++]
+----
+static constexpr ptrdiff_t max() noexcept;
+----
+_Returns_: The maximum number of threads of execution that can be synchronized
+by any `barrier` with the specified `Scope` and `CompletionFunction`.
+
+[source,c++]
+----
+constexpr explicit barrier(ptrdiff_t expected, CompletionFunction f = CompletionFunction());
+----
+_Effects_: Initializes the barrier with the given expected count and completion
+function, and begins the first phase of the barrier.
```

Comment:
The `CompletionFunction` seems hard to use, especially when synchronizing both work-items and host threads.  Do applications have any guarantee about whether it is executed by a host thread vs. a work-item in such a case?  The allowed operations are quite different in those two cases.

Reply:
I hadn't considered that.  C++20 just says that "a thread" executes the completion function during its call to `arrive`, `arrive_and_drop` or `wait`, so I don't think we want to introduce any additional guarantees about which thread executes the function.

What if we add some wording like:

> The completion function may be executed on any thread during a call to `arrive`, `arrive_and_drop` or `wait`.  If a barrier  is used by a mix of host and device threads, it is a user's responsibility to ensure that the function can be called on either the host or device, and that the function obeys any restrictions outlined in the SYCL specification.

Reply:
Yes, I think we can start out like that and see how it goes.  How about wording like this, though:

> The completion function may be executed on any thread during a call to `arrive`, `arrive_and_drop` or `wait`.  The completion function is considered device code, even when the `barrier` is used to synchronize a mixture of host and device threads, so it must abide by the restrictions for device code defined in the core SYCL specification.

This seems stronger than saying "it is the user's responsibility ..." because it makes it clear that an implementation can diagnose errors in the completion function if it uses C++ features that are forbidden in device code.

Reply:
Tweaked as suggested in 4c4c1d7d44f4.